<a href="https://colab.research.google.com/github/akivacooper/ArtificialIntelligence/blob/main/knn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import the libraries math (for square root and absolute value) and pandas and numpy for reading csv files.

In [34]:
import math
import pandas as pd
import numpy as np

Make an untagged vector, point, and two tagged vectors, data1 and data2:

In [2]:
point = [1, 0, 0, '?'] #(an unknown tag)
data1 = [1, 1, 1, 'M']
data2 = [1, 3, 0, 'F']


Write code to separate the data (X) from the tag (Y).  Your output should be:

The vector [1, 1, 1] has tag M

The vector [1, 3, 0] has tag F


In [3]:
def print_vector(array):
  print("The vector " , array[0:-1], " has tag ", array[-1])


In [4]:
print("The vector " , data1[0:-1], " has tag ", data1[-1])
print("The vector " , data2[0:-1], " has tag ", data2[-1])

The vector  [1, 1, 1]  has tag  M
The vector  [1, 3, 0]  has tag  F


Now let's classify the point as either M or F.  We'll do this by setting k = 1 and using the Euclidean distance.  We'll define that as: 

In [49]:
def euclideanDistance(instance1, instance2, length):
   distance = 0
   for x in range(length):
         #print ('x is ' , x)
         num1=float(instance1[x])
         num2=float(instance2[x])
         distance = distance + pow(num1-num2, 2)
   return math.sqrt(distance)


Now, find out the distance between data1 and point and data2 and point. Output what the tag should be using euclideanDistance.

In [6]:
print("EucDis: data1 -> point | " + str(euclideanDistance(data1,point,3)))
print("EucDis: data2 -> point | " + str(euclideanDistance(data2,point,3)))


EucDis: data1 -> point | 1.4142135623730951
EucDis: data2 -> point | 3.0


Now, let's get more data from a file, myFile.csv. This code will read it for us:

In [7]:
url = 'https://github.com/rosenfa/ai/blob/master/myFile.csv?raw=true'
df=pd.read_csv(url,  header=0, error_bad_lines=False) 
#put data in dataset without header line
dataset = np.array(df)

/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py:3326: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


Show that you understand what you read by:

1. Printing the first two vectors in the file
2. Printing the Euclidean distance between those two vectors

Here is some code which I think might help get you on your way!

In [8]:
print(len(dataset))
print(euclideanDistance(dataset[0],dataset[1],3))

10
5.744562646538029


In [9]:
print("first vector: ")
print_vector(dataset[0])
print()
print("second vector: ")
print_vector(dataset[1])
print()
print("distance: " + str(euclideanDistance(dataset[0], dataset[1],3)))

first vector: 
The vector  [0 1 2]  has tag  F

second vector: 
The vector  [1 5 6]  has tag  F

distance: 5.744562646538029


Now assume you have a new value for point:

[0,0,100]

How would you classify this vector using the Euclidean distance function given all of the vectors in the file?

In order to help here is a hint:  

We suggest defining some type of data struction to store different vectors' distances and their tags like this:

In [12]:
point = [0, 0, 100, '?']

In [13]:
class distClass:
    dist = -1 #distance of current point from test point
    tag = '-' #tag of current point


You can then add vector distances like this:

In [14]:
eucDistances = [] # list of distances, will hold objects of type distClass

temp=dataset[1]
label=temp[-1]
d=euclideanDistance(point,temp,3)
print("The distances between " , point , " and " , temp,  " is " , str(d))
print(" and the label is " + label)
obj = distClass() #one record's distance and tag
obj.dist=d
obj.tag=label
eucDistances.append(obj)


The distances between  [0, 0, 100, '?']  and  [1 5 6 'F']  is  94.13819628609845
 and the label is F


In [15]:
eucDistances[0].dist

94.13819628609845

and sort them like this:

In [16]:
eucDistances.sort(key=lambda x: x.dist) 

Questions:

1. What is the label for point if k=1?
2. What is the label for point if k=3?
3. Would the result be different if we used a different distance function like Hamming or Manhattan?

In [17]:
def manhattanDistance(instance1, instance2, length):
   distance = 0
   for x in range(length):
         #print ('x is ' , x)
         num1=float(instance1[x])
         num2=float(instance2[x])
         distance = distance + abs(num1 - num2)
   return distance


def hammingDistance(instance1, instance2, length):
   distance = 0
   for x in range(length):
         #print ('x is ' , x)
         num1=float(instance1[x])
         num2=float(instance2[x])
         distance += int(num1 != num2)
   return distance

In [35]:
def getEucDistances(trainData, unknownPoint, length):
  myEucDistances = []
  myEucDistances.clear()
  for x in range(len(trainData)):
    temp=trainData[x]
    label=temp[-1]
    d=euclideanDistance(unknownPoint,temp,length)
    # print("The distances between " , point , " and " , temp,  " is " , str(d))
    # print(" and the label is " + label)
    obj = distClass() #one record's distance and tag
    obj.dist=d
    obj.tag=label
    myEucDistances.append(obj)
  return myEucDistances

def getHamDistances(trainData, unknownPoint,length):
  myHamDistances = []
  myHamDistances.clear()
  for x in range(len(trainData)):
    temp=trainData[x]
    label=temp[-1]
    d=hammingDistance(unknownPoint,temp,length)
    # print("The distances between " , point , " and " , temp,  " is " , str(d))
    # print(" and the label is " + label)
    obj = distClass() #one record's distance and tag
    obj.dist=d
    obj.tag=label
    myHamDistances.append(obj)
  return myHamDistances

def getManDistances(trainData, unknownPoint,length):
  myManDistances = []
  myManDistances.clear()
  for x in range(len(trainData)):
    temp=trainData[x]
    label=temp[-1]
    d=manhattanDistance(unknownPoint,temp,length)
    # print("The distances between " , point , " and " , temp,  " is " , str(d))
    # print(" and the label is " + label)
    obj = distClass() #one record's distance and tag
    obj.dist=d
    obj.tag=label
    myManDistances.append(obj)  
  return myManDistances

In [36]:
def getKClosest(k, distances):
  
  distances.sort(key=lambda x: x.dist) 
  
  firstSeenTag = distances[0].tag
  secondSeenTag = ''
  firstSeenTagCount = 0
  secondSeenTagCount = 0
  
  for x in range(k):
    curTag = distances[x].tag
    #print("current tag: " + curTag)
    if (curTag == firstSeenTag):
      firstSeenTagCount += 1
    else:
      secondSeenTag = curTag
      secondSeenTagCount += 1
  
  if (firstSeenTagCount > secondSeenTagCount):
    return firstSeenTag
  return secondSeenTag

In [37]:
#Add code with functions that implement Hamming and Manhattan distances and test 
#what the label will be for k=1 and k=3 for all possibilities 
#(6 total: 2 Euclidean, 2 Hamming and 2 Manhattan)

In [38]:
myEucDistances = getEucDistances(dataset, point,3)
myHamDistances = getManDistances(dataset, point,3)
myManDistances = getHamDistances(dataset, point,3)


euc1 = getKClosest(1, myEucDistances)
euc3 = getKClosest(3, myEucDistances)

ham1 = getKClosest(1, myHamDistances)
ham3 = getKClosest(3, myHamDistances)

man1 = getKClosest(1, myManDistances)
man3 = getKClosest(3, myManDistances)


print("euclidean distance | k = 1 | " + euc1);
print("euclidean distance | k = 3 | " + euc3);
print()

print("hamming distance | k = 1 | " + ham1);
print("hamming distance | k = 3 | " + ham3);
print()

print("manhattan distance | k = 1 | " + man1);
print("manhattan distance | k = 3 | " + man3);
print()


euclidean distance | k = 1 | M
euclidean distance | k = 3 | M

hamming distance | k = 1 | M
hamming distance | k = 3 | F

manhattan distance | k = 1 | F
manhattan distance | k = 3 | F



Up until here is a simplified version of the homework.
Below here is the work for the part we will be checking as the basis of your grade: 

Now let's look at some bigger files: 

1. mytrain.csv (for training the model)
2. mytest.csv (for testing)


In [39]:
url = 'https://github.com/rosenfa/ai/blob/master/mytrain.csv?raw=true'
train_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))
url = 'https://github.com/rosenfa/ai/blob/master/mytest.csv?raw=true'
test_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))

/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py:3326: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


In [40]:
print(train_data.shape)# number of records and features
print(train_data)

(100, 30)
[[0.317251 0.540016 0.150514 ... 0.0 0.212007 'F']
 [3.882181 0.0 0.0 ... 0.0 0.0 'F']
 [1.790708 0.0 0.0 ... 0.0 0.0 'F']
 ...
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'F']]


In [41]:
print(test_data.shape)# number of records and features
print(test_data)

(100, 30)
[[2.85985 0.66275 0.0 ... 0.995289 0.0 'F']
 [0.0 12.998726 0.0 ... 0.0 0.0 'F']
 [13.231318 0.0 0.0 ... 0.0 0.0 'F']
 ...
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [1.688611 0.0 3.204529 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'M']]


I hope by now you understand where we are going with this :)

Now implement the knn code with 3 different values for k:
1. k = 1
2. k = 7
3. k = 15

and at first use the Euclidean distance.
Classify each of the vectors in the test_data dataset using the training data from train_data.  Which value for k did the best?  What accuracy did it give you?

Now see if using Hamming or Manhattan distance give any better results for the same values of k.  

Once you are done, you should have a total of 9 different results:

1. Three results for the different value of k using the Euclidean Distance
2. Three results for the different value of k using the Hamming Distance
3. Three results for the different value of k using the Manhattan Distance

Hint: I strongly suggest you base yourself on the code you've seen until this point so you don't have to reinvent the wheel!

In [52]:
#Add code here

kOneCorrect = 0
kSevenCorrect = 0
kFifteenCorrect = 0

for x in range(len(test_data)):

  unknownPoint = test_data[x]
  correctFlag = unknownPoint[-1]

  myEucDistances = getEucDistances(train_data, unknownPoint, 29)

  if getKClosest(1, myEucDistances) == correctFlag:
    kOneCorrect += 1
  if getKClosest(7, myEucDistances) == correctFlag:
    kSevenCorrect += 1
  if getKClosest(15, myEucDistances) == correctFlag:
    kFifteenCorrect +=1

print("Euclidean Results")
print("accuracy for k = 1 | " + str(kOneCorrect) + "%")
print("accuracy for k = 7 | " + str(kSevenCorrect)+ "%")
print("accuracy for k = 15 | " + str(kFifteenCorrect)+ "%")
print()

kOneCorrect = 0
kSevenCorrect = 0
kFifteenCorrect = 0

for x in range(len(test_data)):

  unknownPoint = test_data[x]
  correctFlag = unknownPoint[-1]

  myHamDistances = getHamDistances(train_data, unknownPoint, 29)

  if getKClosest(1, myHamDistances) == correctFlag:
    kOneCorrect += 1
  if getKClosest(7, myHamDistances) == correctFlag:
    kSevenCorrect += 1
  if getKClosest(15, myHamDistances) == correctFlag:
    kFifteenCorrect +=1

print("Hamming Results")
print("accuracy for k = 1 | " + str(kOneCorrect) + "%")
print("accuracy for k = 7 | " + str(kSevenCorrect)+ "%")
print("accuracy for k = 15 | " + str(kFifteenCorrect)+ "%")
print()

kOneCorrect = 0
kSevenCorrect = 0
kFifteenCorrect = 0

for x in range(len(test_data)):

  unknownPoint = test_data[x]
  correctFlag = unknownPoint[-1]

  myManDistances = getManDistances(train_data, unknownPoint, 29)

  if getKClosest(1, myManDistances) == correctFlag:
    kOneCorrect += 1
  if getKClosest(7, myManDistances) == correctFlag:
    kSevenCorrect += 1
  if getKClosest(15, myManDistances) == correctFlag:
    kFifteenCorrect +=1

print("Manhattan Results")
print("accuracy for k = 1 | " + str(kOneCorrect) + "%")
print("accuracy for k = 7 | " + str(kSevenCorrect)+ "%")
print("accuracy for k = 15 | " + str(kFifteenCorrect)+ "%")
print()

Euclidean Results
accuracy for k = 1 | 50%
accuracy for k = 7 | 74%
accuracy for k = 15 | 70%

Hamming Results
accuracy for k = 1 | 61%
accuracy for k = 7 | 55%
accuracy for k = 15 | 57%

Manhattan Results
accuracy for k = 1 | 61%
accuracy for k = 7 | 63%
accuracy for k = 15 | 69%



Which value for k did the best? 
1. Euclidean, k = 7
2. Hamming, k = 1
3. Manhattan, k = 15

What accuracy did it give you?
1. Euclidean, k = 7, accuracy = 74%
2. Hamming, k = 1, accuracy = 61%
3. Manhattan, k = 15, accuracy = 69%

Grade Basis:

80% for correct answers (and yes, there are possibilities that multiple answers are correct-- especially for cases of ties).

20% : Documentation and easily readable code

Please publish your final Notebook in your Github directory.

The homework assignment is due by November 30th.
